packages/interim/src/renovation/component_wrapper/common/template_wrapper.ts

```diff
@@ -2,86 +2,149 @@ import { InfernoComponent, InfernoEffect } from '@devextreme/runtime/inferno';
+// eslint-disable-next-line import/named
+import $, { dxElementWrapper } from '../../../core/renderer';
 import domAdapter from '../../../core/dom_adapter';
-import { getPublicElement } from '../../../core/element';
-import { removeDifferentElements } from '../utils/utils';
+// eslint-disable-next-line import/named
+import { DxElement, getPublicElement } from '../../../core/element';
 import { FunctionTemplate } from '../../../core/templates/function_template';
-import { EffectReturn } from '../../utils/effect_return';
 import { isDefined } from '../../../core/utils/type';
+import { recordMutations } from './mutations_recording';
+
+// eslint-disable-next-line @typescript-eslint/no-type-alias
+type UnknownRecord = Record<PropertyKey, unknown>;
+
+type EqualityComparer = (a?: UnknownRecord, b?: UnknownRecord) => boolean;
 
 export interface TemplateModel {
-  data: Record<string, unknown>;
+  data: UnknownRecord & {
+    isEqual?: EqualityComparer;
+  };
   index: number;
 }
 
-interface TemplateWrapperProps {
+export interface TemplateWrapperProps {
   template: FunctionTemplate;
   model?: TemplateModel;
+  isEqual?: EqualityComparer;
   transclude?: boolean;
   renovated?: boolean;
 }
 
+function isDxElementWrapper(
+  element: dxElementWrapper | HTMLElement & Partial<Pick<dxElementWrapper, 'toArray'>>,
+): element is dxElementWrapper {
+  return !!element.toArray;
+}
+
+type TemplateModelArgs =
+  // eslint-disable-next-line @typescript-eslint/no-type-alias
+  Required<Pick<TemplateWrapperProps, 'model'>>
+  // eslint-disable-next-line @typescript-eslint/no-type-alias
+  & Omit<TemplateWrapperProps, 'model'>;
+
+export function buildTemplateArgs(
+  model: TemplateModel,
+  template: TemplateWrapperProps['template'],
+): TemplateModelArgs {
+  const args: TemplateModelArgs = {
+    template,
+    model: { ...model },
+  };
+
+  const { isEqual, ...data } = model.data ?? {};
+  if (isEqual) {
+    args.model.data = data;
+    args.isEqual = isEqual;
+  }
+
+  return args;
+}
+
+function buildTemplateContent(
+  props: TemplateWrapperProps,
+  container: DxElement<Element>,
+): ChildNode[] {
+  const {
+    data, index,
+  } = props.model ?? { data: {} };
+
+  if (data) {
+    Object.keys(data).forEach((name) => {
+      if (data[name] && domAdapter.isNode(data[name])) {
+        data[name] = getPublicElement($(data[name] as Element));
+      }
+    });
+  }
+
+  const rendered = props.template.render({
+    container,
+    transclude: props.transclude,
+    ...{ renovated: props.renovated },
+    ...!props.transclude ? { model: data } : {},
+    ...!props.transclude && Number.isFinite(index) ? { index } : {},
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  }) as any as (dxElementWrapper | DxElement | undefined);
+
+  if (rendered === undefined) {
+    return [];
+  }
+
+  return isDxElementWrapper(rendered)
+    ? rendered.toArray()
+    : [$(rendered).get(0)];
+}
+
+function noop(): void {}
```

Comment:
Minor: noop is already exists in core lib

Reply:
done, thanks